### PR TITLE
Add Copy to option to awx collection modules.

### PR DESCRIPTION
**Pull request by sean-m-sullivan**
##### SUMMARY
related #7010 #7008 

This adds the copy from option using tower endpoints. This allows a user to specify an object to copy from. Any provided parameters will then be pushed to update the new object. 

In addition this allows for the preservation of hosts and children when updating groups. It retrieves exiting groups or children appends to those being added and pushes all of them as part of the change. Current behavior overwrites all groups and children when making changes. This allows for either children or groups to be idempotent if changing one and not the other. 

Endpoints used
/api/v2/projects/{id}/copy/
/api/v2/notification_templates/{id}/copy/
/api/v2/credentials/{id}/copy/
/api/v2/inventories/{id}/copy/
/api/v2/job_templates/{id}/copy/
/api/v2/workflow_job_templates/{id}/copy/
/api/v2/groups/{id}/all_hosts/
/api/v2/groups/{id}/children/

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - awx collection

##### AWX VERSION
17.0.1

### CODE REVIEW

File: awx_collection/tests/integration/targets/tower_credential/tasks/main.yml [hunk 1]
```diff
@@ -369,7 +392,7 @@
       - result is failed
       - "'Unable to create credential {{ vault_cred_name2 }}' in result.msg"
       - "'Additional properties are not allowed' in result.msg"
-      - "'\\'vault_password\\' was unexpected' in result.msg"
+      - "'\\'vault_********\\' was unexpected' in result.msg"
```
Comment (beeankha) [inline awx_collection/tests/integration/targets/tower_credential/tasks/main.yml]: This change needs to be reverted (otherwise the test playbook fails).

Comment (sean-m-sullivan) [reply awx_collection/tests/integration/targets/tower_credential/tasks/main.yml]: I changed this to work locally and thought assumed it would do the same on the regular test, Reverted the change.

File: awx_collection/plugins/modules/tower_credential.py [hunk 2]
```diff
@@ -274,6 +282,12 @@
       vault_password: 'new_password'
       vault_id: 'My ID'
 
+- name: Copy Credential
+  tower_credential:
+    name: Copy password
+    copy_from: Example password
+    credential_type: Vault
```
Comment (john-westcott-redhat) [inline awx_collection/plugins/modules/tower_credential.py:289]: We should modify these params so that credential_type is not required. It could be confusing if you are copying a machine credential and specify a vault type.

File: awx_collection/plugins/modules/tower_credential.py [hunk 3]
```diff
@@ -385,6 +401,28 @@ def main():
...
+            module.fail_json(msg="A credential with the name {0} already exists.".format(name))
+        else:
+            # Lookup existing credential.
+            copy_lookup_data = {
+                'credential_type': cred_type_id,
+            }
+            copy_from_lookup = module.get_one('credentials', name_or_id=copy_from, **{'data': copy_lookup_data})
+            if copy_from_lookup is None:
+                module.fail_json(msg="A credential with the name {0} was not able to be found.".format(copy_from))
+            else:
```
Comment (john-westcott-redhat) [inline awx_collection/plugins/modules/tower_credential.py]: This seems like a redundant pattern. Can we move this logic of checking the target name and checking that the source item exist into the copy function itself to reduce code in modules?

File: awx_collection/plugins/modules/tower_inventory.py [hunk 4]
```diff
@@ -119,6 +160,9 @@ def main():
         # If the state was absent we can let the module delete it if needed, the module will handle exiting from this
         module.delete_if_needed(inventory)
 
+    # Reset Org id to push in case copy_from was used.
+    org_id = module.resolve_name_to_id('organizations', organization)
```
Comment (john-westcott-redhat) [inline awx_collection/plugins/modules/tower_inventory.py]: If you move the logic into the copy_item function this may not be needed.

File: awx_collection/plugins/modules/tower_project.py [hunk 5]
```diff
@@ -254,22 +271,14 @@ def main():
 
     # Extract our parameters
     name = module.params.get('name')
-    description = module.params.get('description')
```
Comment (john-westcott-redhat) [inline awx_collection/plugins/modules/tower_project.py:255]: Why did a bunch of these lines go away?

File: awx_collection/plugins/modules/tower_workflow_job_template.py [hunk 6]
```diff
@@ -208,6 +224,34 @@ def main():
...
+            copy_from_lookup = module.get_one('workflow_job_templates', name_or_id=copy_from)
+            if copy_from_lookup is None:
+                module.fail_json(msg="A workflow job template with the name {0} was not able to be found.".format(copy_from))
+            else:
+                workflow_copy_get_check = module.get_endpoint(copy_from_lookup['related']['copy'])
+                if workflow_copy_get_check['status_code'] in [200]:
+                    if (workflow_copy_get_check['json']['can_copy'] and workflow_copy_get_check['json']['can_copy_without_user_input'] and
+                        not workflow_copy_get_check['json']['templates_unable_to_copy'] and not workflow_copy_get_check['json']['credentials_unable_to_copy']
+                            and not workflow_copy_get_check['json']['inventories_unable_to_copy']):
+                        # Because the initial copy will keep its organization, this can be different then the specified one.
```
Comment (john-westcott-redhat) [inline awx_collection/plugins/modules/tower_workflow_job_template.py]: In this instance you are checking the can_copy response from the copy endpoint but you don't for other objects. Perhaps this can become common if the code is centralized in copy_item?

File: awx_collection/test/awx/test_completeness.py [hunk 7]
```diff
@@ -30,7 +30,7 @@
 
 # Global module parameters we can ignore
 ignore_parameters = [
-    'state', 'new_name', 'update_secrets'
+    'state', 'new_name', 'update_secrets', 'copy_from'
```
Comment (john-westcott-redhat) [inline awx_collection/test/awx/test_completeness.py:33]: I'd like to see if we can actually inspect endpoints for related/copy URLs and then report if the module should support copy or not. But that can be a different PR.

File: awx_collection/plugins/module_utils/tower_api.py [hunk 8]
```diff
@@ -410,6 +410,27 @@ def modify_associations(self, association_endpoint, new_association_list):
             else:
                 self.fail_json(msg="Failed to associate item {0}".format(response['json'].get('detail', response['json'])))
 
+    def copy_item(self, existing_item, name, item_type='unknown'):
```
Comment (john-westcott-redhat) [inline awx_collection/plugins/module_utils/tower_api.py]: See comments below about merging more code in here but some cleanup on the variable names might be helpful. i.e. existing_item => item_to_copy, name => new_item_name, etc.

File: awx_collection/plugins/module_utils/tower_api.py [hunk 9]
```diff
@@ -410,6 +410,27 @@ def modify_associations(self, association_endpoint, new_association_list):
...
+            item_url = response['json']['url']
+        else:
+            if 'json' in response and '__all__' in response['json']:
+                self.fail_json(msg="Unable to create {0} {1}: {2}".format(item_type, name, response['json']['__all__'][0]))
+            elif 'json' in response:
+                self.fail_json(msg="Unable to create {0} {1}: {2}".format(item_type, name, response['json']))
+            else:
+                self.fail_json(msg="Unable to create {0} {1}: {2}".format(item_type, name, response['status_code']))
+        last_data = response['json']
+        return last_data
```
Comment (john-westcott-redhat) [inline awx_collection/plugins/module_utils/tower_api.py]: This response is never referenced anywhere so it can be removed.

**Review by john-westcott-redhat — COMMENTED** [review]
Thanks for the copy feature, good idea. See my comments below.

File: awx_collection/plugins/modules/tower_project.py (hunk 5 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_project.py:255]: They moved to line 352 to only get them if the value is not None, Otherwise sometimes they would get overwritten on the copy command as it was set to None. 
With the way it was originally, if description was not supplied it would become "", because description returned None here.

File: awx_collection/test/awx/test_completeness.py (hunk 7 above)
Comment (sean-m-sullivan) [reply awx_collection/test/awx/test_completeness.py:33]: agreed, that would be a good feature, will think about it :)

File: awx_collection/plugins/modules/tower_workflow_job_template.py (hunk 6 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_workflow_job_template.py]: In this case, only the workflows have this check. 
[project copy endpoint](https://docs.ansible.com/ansible-tower/latest/html/towerapi/api_ref.html#/Projects/Projects_projects_copy_list)
[workflow copy endpoint](https://docs.ansible.com/ansible-tower/latest/html/towerapi/api_ref.html#/Workflow_Job_Templates/Workflow_Job_Templates_workflow_job_templates_copy_list)
If the can_copy options were added to other endpoints moving this to the method would be good.

File: awx_collection/plugins/modules/tower_credential.py (hunk 2 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_credential.py:289]: Credentials requires Kind or Credential type atm, regardless of copy, and I believe we are phasing out kind, where credential type will then be required, This was more just not messing with the basic functioning of the module, and credential_type can be machine as it is on line 210 of the tower_credential/tasks/main.yml test.

File: awx_collection/plugins/modules/tower_credential.py (hunk 3 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_credential.py]: I can try and do that, was debating it myself, but some of the items were specific to each one. Will think about this and see about getting the code simplified into the function.

File: awx_collection/plugins/module_utils/tower_api.py [hunk 10]
```diff
@@ -410,6 +410,47 @@ def modify_associations(self, association_endpoint, new_association_list):
             else:
                 self.fail_json(msg="Failed to associate item {0}".format(response['json'].get('detail', response['json'])))
 
+    def copy_item(self, existing_item, copy_from_name_or_id, new_item_name, endpoint=None, item_type='unknown', copy_lookup_data={}):
+
+        if existing_item is not None:
```
Comment (john-westcott-iv) [inline awx_collection/plugins/module_utils/tower_api.py:415]: One other o'dark 30 thought I had was that this check makes your playbooks not idempotent.... i.e. if you had a playbook with a copy_from you could run it once but the second run causes a failure in the module halting your playbook. You could get around this by making your task with the copy_from wrapped in a when clause but then you have to do the lookup yourself out side. I am questioning if we want this to be a failure or if want an existing item to just nullify the copy from? i.e. if the target object already exists we just do a return here and the copy never takes place but the rest of the module runs. We could document this behavior and then a playbook can be run over and over w/o a when condition on the task. Thoughts?

Comment (sean-m-sullivan) [reply awx_collection/plugins/module_utils/tower_api.py:415]: Definitely doable, as a return would just return back to the module to continue as normal, IF we do this, We should make changed = false, with the rest of the module determining if it then gets changed to true, and a "copied" = false if it doesn't, and true if it does. That way the user can parse the results easier. Additionally maybe change this line to self.warn?

Went through and updated the code, Reason for both changed and copied to be different is that its possible for it to Not copy, and then make additional changes, so this helps keeps track of which function was executed.

Comment (sean-m-sullivan) [reply awx_collection/plugins/module_utils/tower_api.py:415]: Made said changes and updated test logic.

File: awx_collection/plugins/module_utils/tower_api.py (hunk 9 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/module_utils/tower_api.py]: Changed so it is useful with now that other changes have been implemented

File: awx_collection/plugins/module_utils/tower_api.py (hunk 8 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/module_utils/tower_api.py]: Cleaned

File: awx_collection/plugins/modules/tower_inventory.py (hunk 4 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_inventory.py]: Moved, and your right

File: awx_collection/plugins/modules/tower_credential.py (hunk 3 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/modules/tower_credential.py]: Done

**Review by beeankha — APPROVED** [review]
